apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -308,14 +293,22 @@ choose(N, List) ->
     List2 = [X || {_, X} <- List1],
     lists:sublist(List2, N).
 
--spec(upstream_failures([upstream()]) -> non_neg_integer()).
+-spec(upstream_failures(upstream()) -> number()).
 upstream_failures(Upstream) ->
-    case exometer:get_value([?MODULE, Upstream, failures]) of
-        {error, _Error} ->
-            0;
-        {ok, Metric} ->
-            {one, Failures} = lists:keyfind(one, 1, Metric),
-            Failures
+    UpstreamFailuresFilterMap = upstream_failures_filtermap_fun(Upstream),
+    AllFailures = prometheus_counter:values(dns, worker_failures_total),
+    UpstreamFailures = lists:filtermap(UpstreamFailuresFilterMap, AllFailures),
+    lists:sum(UpstreamFailures).
+
+-spec(upstream_failures_filtermap_fun(upstream()) ->
+   fun(({nonempty_maybe_improper_list(), number()}) -> false | {true, number()})).
+upstream_failures_filtermap_fun(Upstream) ->
+    QueryAddress = upstream_to_binary(Upstream),
+    fun({[{"upstream_address", Address} | _], Failures}) ->
```

Comment:
It's broken here.

Reply:
That is interesting. I am pretty sure that worked before, wit the same custom resolvers we used on my firs PR. I think at the end of the last PR I changed the order of the labels, so upstream address is not the first anymore `{labels, [protocol, upstream_address, reason]}`. That would have matched and then this error would not have happened.

I agree though that it is more robust to write code that does not depend on the order of the labels, as I already made a change that broke it :).